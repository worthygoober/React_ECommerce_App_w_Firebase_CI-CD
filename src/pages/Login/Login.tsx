import { useState, useEffect } from "react";
import { signInWithEmailAndPassword} from "firebase/auth";
import { auth } from "../../lib/firebase/firebase";
import { useAuth } from "../../context/AuthContext";
import { useNavigate, Link } from "react-router-dom";
import './Login.css';
import { Flip, ToastContainer, toast } from "react-toastify";


const Login: React.FC = () => {
    const [email, setEmail] = useState<string>('');
    const [password, setPassword] = useState<string>('');

    const { user } = useAuth();

    const navigate = useNavigate();

    // redirect user to profile page after logging in
    useEffect(() => {
        if (user) {
            navigate('/profile')
        }
    }, [user, navigate]);

    if (user) {
        return null;
    };
    
    // login function using firebase functionality and auth state
    const handleLogin = async (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        try {
            await signInWithEmailAndPassword(
                auth,
                email,
                password
            );
            toast.success("Login successful!");
            navigate('/profile');
        } catch (err: any) {
            toast.error(err.message, {
                autoClose: 5000,
            });
            console.error(err.message);
        }
    };

    return (
    <>
        <div className="form">
            <h1>Welcome back</h1>
            <h2>Please Login</h2>

            {/* login form */}
            <form onSubmit={handleLogin}>
                <fieldset className="fieldset">
                <legend className="legend">Login</legend>
                    
                    <input
                    className="input"
                    type="email"
                    placeholder="Email"
                    value={email}
                    onChange={(e) => setEmail(e.target.value)}
                    />
                    
                    <input
                    className="input"
                    type="password"
                    placeholder="Password"
                    value={password}
                    onChange={(e) => setPassword(e.target.value)}
                    />

                </fieldset>

                <button type="submit" className="button">Login</button>

                {/* link/button that takes new users to register page */}
                <Link to='/register' className="button">New Users</Link>
            </form>

            {/* styling for success/error messages */}
            <ToastContainer 
                position="top-center"
                autoClose={3000}
                hideProgressBar={false}
                newestOnTop={false}
                closeOnClick={false}
                rtl={false}
                pauseOnFocusLoss
                draggable
                pauseOnHover
                theme="dark"
                transition={Flip}
            />
        </div>
    </>
  );
};

export default Login;