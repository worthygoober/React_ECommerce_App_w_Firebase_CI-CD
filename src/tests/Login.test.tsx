// Snapshot testing

import { render } from '@testing-library/react';
import Login from '../pages/Login/Login';

// 1. Mock Firebase
// jest.mock('../lib/firebase.ts', () => ({
//   auth: {
//     signInWithEmailAndPassword: jest.fn(() => Promise.resolve({ 
//       user: {
//         uid: 'test-uid',
//         email: 'test@example.com'
//       } 
//     })),
//     // Add other auth methods if used in component
//     onAuthStateChanged: jest.fn(),
//     signOut: jest.fn()
//   },
//   firebaseConfig: {
//     apiKey: 'test-api-key',
//     authDomain: 'test.firebaseapp.com',
//     projectId: 'test-project',
//     storageBucket: 'test.appspot.com',
//     messagingSenderId: '123456789',
//     appId: '1:123456789:web:abcdef123456'
//   }
// }));

jest.mock('firebase/auth', () => ({
    getAuth: jest.fn(() => ({
      signInWithEmailAndPassword: jest.fn(() => Promise.resolve({
        user: {
          uid: 'test-uid',
          email: 'test@example.com'
        }
      })),
      onAuthStateChanged: jest.fn(),
      signOut: jest.fn(),
      currentUser: null
    }))
}));
  
jest.mock('firebase/app', () => ({
    initializeApp: jest.fn().mockReturnValue({}),
}));

jest.mock('firebase/firestore', () => ({
    getFirestore: jest.fn(() => ({})),
}));

// 2. Mock environment variables
beforeAll(() => {
    // Mock all Vite environment variables used in firebase.ts
    process.env = {
      ...process.env, // Preserve existing env vars
      VITE_APP_FIREBASE_API_KEY: 'test-api-key-mock',
      VITE_APP_FIREBASE_AUTH_DOMAIN: 'mock-project.firebaseapp.com',
      VITE_APP_FIREBASE_PROJECT_ID: 'mock-project-id',
      VITE_APP_FIREBASE_STORAGE_BUCKET: 'mock-project.appspot.com',
      VITE_APP_FIREBASE_MESSAGING_SENDER_ID: '1234567890',
      VITE_APP_FIREBASE_APP_ID: '1:1234567890:web:mockappid',
      
    //   // Add any other Vite env vars your app uses
    //   VITE_APP_OTHER_SERVICE_KEY: 'other-service-key-mock'
    };
  });


// 3. Mock react-router-dom
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => jest.fn(),
  Link: ({ children }: { children: React.ReactNode }) => children,
}));

// 4. Mock AuthContext if used
jest.mock('../context/AuthContext', () => ({
  useAuth: () => ({
    currentUser: null,
    loading: false,
    login: jest.fn(),
    logout: jest.fn()
  }),
}));

// Cleanup after each test
afterEach(() => {
  jest.clearAllMocks();
  // Clean up any rendered components
//   document.body.innerHTML = '';
});

describe('Login Component', () => {
  test('matches snapshot', () => {
    const { asFragment } = render(<Login />);
    expect(asFragment()).toMatchSnapshot();
  });
});