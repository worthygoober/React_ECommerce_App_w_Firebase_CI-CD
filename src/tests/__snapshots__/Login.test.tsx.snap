// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="form"
  >
    <h1>
      Welcome back
    </h1>
    <h2>
      Please Login
    </h2>
    <form>
      <fieldset
        class="fieldset"
      >
        <legend
          class="legend"
        >
          Login
        </legend>
        <input
          class="input"
          placeholder="Email"
          type="email"
          value=""
        />
        <input
          class="input"
          placeholder="Password"
          type="password"
          value=""
        />
      </fieldset>
      <button
        class="button"
        type="submit"
      >
        Login
      </button>
      New Users
    </form>
    <section
      aria-atomic="false"
      aria-label="Notifications Alt+T"
      aria-live="polite"
      aria-relevant="additions text"
      class="Toastify"
    />
  </div>
</DocumentFragment>
`;
